2024-10-30-lapply/cc_lapply.Rmd

```diff
@@ -0,0 +1,694 @@
+
+Lets say we required the total items, cost and distinct chemical substances, per practice, from this data.
+Lets also say, we also wanted this aggregated done per practice separately (for a reporting reason of some kind).
+Instead of writing 8k+ chunks of code, it would obviously make more sense to write a function to do this.
+Something along the lines of
+
+```{r}
+# Function to sum metrics by practice
+sum_items_by_practice = function(df, practice_code){
+  
+  # Generate output
+  output = df %>% 
+    filter(PRACTICE_CODE == practice_code) %>% 
+    group_by(
+      REGIONAL_OFFICE_NAME,
+      PRACTICE_NAME, 
+      PRACTICE_CODE
+      ) %>% 
+    summarise(
+      ITEMS = sum(ITEMS),
+      NIC = sum(NIC),
+      DRUG_COUNT = n_distinct(CHEMICAL_SUBSTANCE_BNF_DESCR),
+      .groups = "drop"
+      )
+  
+  # Return
+  return(output)
+}
+```
+
+We can check it with a single practice code:
+
+```{r}
+sum_items_by_practice(df, 'Y05949')
+```
+
+Using lapply we can map a function to a vector. This vector could be:
+
++ Integers
++ Characters
+
+If we can create a vector of practice codes, we can use this with the above function.
+
+```{r}
+# Generate vector of practice names
+practice_codes_vec = df %>% 
+  select(PRACTICE_CODE) %>% 
+  distinct() %>% 
+  pull()
+
+# Check practice codes
+head(practice_codes_vec,50)
+
+# Check length
+length(practice_codes_vec)
+```
+
+We can now apply the function to our vector.
+We will time how long it takes using 'tic()' and 'toc()' from the 'tictoc' package.
+If a function has more than input:
+
++ The 'X' parameter is your vector
++ Explicitly state the other parameter inputs
+
+The code is hashed out, but would take ~3.5 minutes to run.
+
+```{r}
+# # Start timer
+# tic()
+# 
+# # Lapply
+# results = lapply(
+#   X = practice_codes_vec,
+#   FUN = sum_items_by_practice,
+#   df = df
+#   )
+# 
+# # End timer
+# toc()
+```
+
+### Run in parLapply
+
+This is a fairly slow process, that could be easily sped up.
+We can use parLapply to run the process in parallel.
+This is a very similar code structure to 'regular' lapply. 
+
+There are a few considerations to bear in mind when using parLapply:
+
++ How many cores your machine has
++ How many cores you want to *not* be used
++ The size of your function inputs/dependencies
++ E.g., if you want x6 processing, and you have a 500MB dataframe, this needs copying x6 times.
+
+The process then has a few steps:
+
+1. Define how many cores you want to use
+2. 'Make' your cluster
+3. Export all the required libraries (used in the function) to each cluster
+4. Export all the required objects (used in the function) to each cluster
+5. Generate the results
+6. Stop the cluster (IMPORTANT!)
+
+The function itself:
+
++ You need to specify the cluster object, in the 'cl' parameter.
++ The 'FUN' parameter becomes 'fun'
+
+Can first choose number of cores and setup cluster.
+
+```{r}
+# Generate appropriate number of cores
+n_cores = detectCores() - 2
```

Comment:
[Interesting reading: Please Avoid detectCores() in your R Packages](https://www.jottr.org/2022/12/05/avoid-detectcores/).

Even tho C&C this is just demoing the functionality, useful to make people aware of potential issues.